src/machine/machine_avr.go

```diff
@@ -271,3 +272,122 @@ func (i2c I2C) ReadByte() byte {
 
 	return byte(*avr.TWDR)
 }
+
+// UART
+
+type UARTConfig struct {
+	Baudrate uint32
+}
+
+type UART struct {
+}
+
+var (
+	// UART0 is the hardware serial port on the AVR.
+	UART0 = &UART{}
+)
+
+// Configure the UART on the AVR.
+func (uart UART) Configure(config UARTConfig) {
+	// default to 115200
+	if config.Baudrate == 0 {
+		config.Baudrate = 115200
+	}
+
+	// Set baud rate based on prescale formula:
+	// BAUD_PRESCALE = (((F_CPU / (BAUDRATE * 16UL))) - 1)
+	ps := (CPU_FREQUENCY / (config.Baudrate * 16)) - 1
+	*avr.UBRR0H = avr.RegValue(ps >> 8)
+	*avr.UBRR0L = avr.RegValue(ps)
+
+	// enable RX interrupt
+	*avr.UCSR0B |= avr.UCSR0B_RXCIE0
+}
+
+// Close the UART on the AVR.
+func (uart UART) Close() error {
+	return nil
+}
+
+// Read from the RX buffer.
+func (uart UART) Read(data []byte) (n int, err error) {
+	if len(data) > bufferSize {
+		return 0, errors.New("Read buffer cannot be larger than RX Buffer")
+	}
+
+	// check if RX buffer is empty
+	size := uart.BufferUsed()
+	if size == 0 {
+		return 0, nil
+	}
+
+	// only read number of bytes used from buffer
+	for i := 0; uint8(i) < size; i++ {
+		data[i] = byte(bufferGet())
+	}
+
+	return len(data), nil
+}
+
+// Write data to the UART.
+func (uart UART) Write(data []byte) (n int, err error) {
+	for _, v := range data {
+		uart.WriteByte(v)
+	}
+	return len(data), nil
+}
+
+// ReadByte reads a single byte from the RX buffer.
+func (uart UART) ReadByte() (byte, error) {
+	return byte(bufferGet()), nil
+}
+
+// WriteByte writes a byte of data to the UART.
+func (uart UART) WriteByte(c byte) error {
+	// Wait until UART buffer is not busy.
+	for (*avr.UCSR0A & avr.UCSR0A_UDRE0) == 0 {
+	}
+	*avr.UDR0 = avr.RegValue(c) // send char
+	return nil
+}
+
+type __volatile byte
+
+const bufferSize = 64
+
+// Minimal ring buffer implementation inspired by post at
+// https://www.embeddedrelated.com/showthread/comp.arch.embedded/77084-1.php
+var rxbuffer [bufferSize]__volatile
+var head __volatile
+var tail __volatile
+
+func (uart UART) BufferUsed() uint8 {
+	return bufferUsed()
+}
+
+func bufferUsed() uint8 { return uint8(head - tail) }
```

Comment:
In the bufio package, they use `Buffered() int` (for both the reader and the writer).
I think it is best to stick to that API wherever it makes sense for an UART. And as it is only really required to have a buffer for incoming bytes due to it being interrupt driven (tx just blocks, rx might miss bytes without an isr/buffer) I think we should use `Buffered() int` here.
What do you think?